oecophylla/distance/distance.rule

```diff
@@ -0,0 +1,178 @@
+rule mash_sketch:
+    """
+    Sketches a quality-controlled fastq file with mash.
+    """
+    input:
+        forward = qc_dir + "{sample}/filtered/{sample}.R1.trimmed.filtered.fastq.gz"
+    output:
+        distance_dir + '{sample}/mash/{sample}.msh'
+    params:
+        env = config['envs']['distance'],
+        size = config['params']['mash']['sketch_size'],
+        m = config['params']['mash']['min_obs'],
+        kmer = config['params']['mash']['kmer']
+    threads:
+        1
+    log:
+        distance_dir + "logs/mash_sketch.sample=[{sample}].log"
+    benchmark:
+        "benchmarks/distance/mash_sketch.sample=[{sample}].txt"
+    run:
+        output_base = os.path.splitext(output[0])[0]
+
+        shell("""
+              set +u; {params.env}; set -u
+
+              mash sketch -r -k {params.kmer} \
+              -m {params.m} -s {params.size} \
+              -o {output_base} {input.forward}
+              """)
+
+rule mash_dist:
+    """
+    compute mash distance between every pair of samples.
+    """
+    input:
+        expand(rules.mash_sketch.output, sample=samples)
+    output:
+        distance_dir + 'combined_analysis/mash.dist.txt'
+    threads:
+        1
+    log:
+        distance_dir + "logs/mash_dist.log"
+    benchmark:
+        "benchmarks/distance/mash_dist.txt"
+    run:
+        from itertools import combinations
+        for i, j in combinations(input, 2):
+            shell('mash dist {i} {j} >> {output[0]}')
+
+
+rule mash_dm:
+    """
+    Make a distance matrix.
+    """
+    input:
+        rules.mash_dist.output
+    output:
+        dist_matrix = distance_dir + 'combined_analysis/mash.dist.dm',
+        p_matrix = distance_dir + 'combined_analysis/mash.dist.p'
+    threads:
+        1
+    log:
+        distance_dir + "logs/mash_dm.log"
+    benchmark:
+        "benchmarks/distance/mash_dm.txt"
+    run:
+        from skbio.stats.distance import DissimilarityMatrix
+        import pandas as pd
+        import numpy as np
+
+        mash_vec = pd.read_csv(input[0], sep = '\t', header=None)
+
+        # get sorted list of samples
+        samples = sorted(set(mash_vec[0]) | set(mash_vec[1]))
+
+        dm = np.zeros([len(samples),len(samples)])
+        pm = np.zeros([len(samples),len(samples)])
+
+        # fill matrices with values
+        for s1, s2, d, p in zip(mash_vec[0],mash_vec[1],mash_vec[2],mash_vec[3]):
```

Comment:
Maybe for the sake of speed, it would be nice make these bulk operations to get rid of the python loop.

One possibility is to first convert this a [scipy sparse matrix](https://docs.scipy.org/doc/scipy/reference/sparse.html) (possibly [coo_format](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.coo_matrix.html#scipy.sparse.coo_matrix))
, and convert that into a `skbio.DistanceMatrix`.

I raised an issue here, so we don't need to worry about this now.

Reply:
Are these actually sparse data? If yes, then they should not represented as a `skbio.DissimilarityMatrix` which is a dense matrix behind the scenes. If no, then the transformation to sparse, only to transform to dense, is a lot of wasted compute; the for loop here is O(n).

Reply:
good to know, but this is dense.